node/src/exchange.rs

```diff
@@ -31,21 +34,78 @@ pub fn new_behaviour(network: &str) -> Behaviour {
 #[derive(Clone, Copy, Debug, Default)]
 pub struct HeaderCodec;
 
+impl HeaderCodec {
+    async fn read_raw_message<T, R>(
+        io: &mut T,
+        buf: &mut Vec<u8>,
+        max_len: usize,
+    ) -> io::Result<Option<R>>
+    where
+        T: AsyncRead + Unpin + Send,
+        R: Message + Default,
+    {
+        let mut read = 0;
+        let len = loop {
+            match io.read(&mut buf[read..]).await? {
+                0 => {
+                    // check if we're between Messages, in which case it's ok to stop
+                    if read == 0 {
+                        return Ok(None);
+                    } else {
+                        return Err(io::Error::new(
+                            io::ErrorKind::UnexpectedEof,
+                            ReadHeaderError::StreamClosed,
+                        ));
+                    }
+                }
+                n => read += n,
+            };
+
+            if let Ok(len) = prost::decode_length_delimiter(&buf[..read]) {
+                break len;
+            }
```

Comment:
Right, this was incorrectly mapped to `StreamClosed` since we were trying to `io.read` to zero length buffer.
Fixed and added `ReadHeaderError::VarintOverflow`.